content/software-testing-the-hardware-software-interface-part-2.md

```diff
@@ -0,0 +1,417 @@
+|   (Controller)     |             |       (DUT)           |
+|                    |             |     [Black Box]       |
+|                    |   poking    |                       |
+|  - Sends Inputs    | --------->  |  - Processes Inputs   |
+|    via UART,       |             |                       |
+|    Display, etc.   |             |                       |
+|                    |   reacting  |                       |
+|  - Receives        | <---------  |  - Generates Outputs  |
+|    Outputs         |             |    via UART, etc.     |
+|    via UART,       |             |                       |
+|    Other I/O       |             |                       |
++--------------------+             +-----------------------+
+```
+
+### Practice
+
+So, as an example let's assume you need to monitor whether your Linux box boots. The manual steps to achieve this are:
+
+1. Reset the board from whatever state it was
+2. Wait for the bootloader, kernel and userspace to boot
+3. See the login prompt
+
+Great, now we've reduced a requirement (the board must boot) to a set of manual execution steps. But you don't have just
+one board, you have around 100 different ones, and every single one of them must be tested every time your Continuous
+Integration builds a new OS for the board. Thinking about it in terms of automation, we can reduce this to:
+
+1. Connect the board's UART to a "Controller" - simply put, hooking up the board to a computer
+2. Write some software that parses the logs from the UART
+3. If it detects our login prompt, consider the boart fully booted
+
+As a practical example, I have a Verdin iMX8MP SoM connected to an Dahlia carrier board. Power is connected as is the
+UART to my Controller, which in this case is my computer.
+
+```
++-------------------+          UART Cable          +------------------------+
+|                   | <------------------------->  |                        |
+|   Controller      |                              |  Dahlia Carrier Board  |
+|   (A Computer)    |                              |  + Verdin iMX8MP SoM   |
+|                   |                              |                        |
++-------------------+                              +------------------------+
+                                                        |
+                                                        |
+                                                   +------------+
+                                                   |  Power     |
+                                                   |  Supply    |
+                                                   +------------+
+```
+
+Then it's a matter of writing some software that runs in the controller and waits for the login prompt.
+
+> You can choose whatever language you feel comfortable with writing tests, but I'd highly suggest looking into either
+Python, Perl or Tcl. Interpreted languages, dynamic are generally better for these tasks because you want to chug out
```

Comment:
"Interpreted languages, dynamic are generally" --> "Interpreted, dynamic languages are generally"

I may be wrong, but I have a feeling that the proper structure is a comma-separated list of adjectives + the noun.

Reply:
Fixed, thanks.